### ISSUE TITLE
About using libostree as a data version control

### ISSUE THREAD

**Issue by adityashah1212** [ostreedev/ostree#1875]
Hi,

I was thinking if it was possible to create a version control system for pure data or to be more accurate to snapshot data in a certain state and reproduce it using libostree. I know this is not why this project was created, but since the underlying idea of deduplication and reproduction of the data would a lot like what we do for OS, I think this could on of use cases. I don't want to fork from this project and maintain it separately if possible.

**Comment by mwleeds** [comment]
I think you could if the data is being stored in files on a filesystem (as opposed to say a database), but why not just use git?

**Comment by adityashah1212** [comment]
Yes it supposed to files, but my purpose here is to snapshot it and then user-mode checkout each commit into a different directory. Of course it is assumed that the whole thing will happen on the same filesystem. The problem with git is that when you checkout the files, it would make a copy of files rather than hard linking them (correct me if I am wrong). This will not scale when the number of commits grow.

**Comment by mwleeds** [comment]
Yeah, based on `ls -l` it seems you're right that git doesn't hard link files to check them out.

**Comment by jlebon** [comment]
> to snapshot data in a certain state and reproduce it using libostree

Hmm, what are you missing right now from libostree to do this? You could write a thin layer on top to make it even more "git-CLI-like" if you'd like (e.g keeping it in a `.ostree` repo or something), but all the basic tools should already be there.

> Yeah, based on `ls -l` it seems you're right that git doesn't hard link files to check them out.

Otherwise, it'd be impossible to hack on a git checkout without corrupting the object store. :)

**Comment by cgwalters** [comment]
> Otherwise, it'd be impossible to hack on a git checkout without corrupting the object store. :)

Not if we had [O_OBJECT](https://marc.info/?l=linux-fsdevel&m=139963046823575&w=2) - git could use it too.
